### PR TITLE
Acids, Smoke, Foam, reagents reaction() refactors

### LINKED ISSUE #6276: You can still put acid in spray bottles using syringes
REMIEEEEEEEEEEEEEEEEEEEEEEEE

### LINKED ISSUE #6857: Poly acid can melt most clothing with a single unit from a dropper.
This is kind of ridiculously powerful.

OOC: DarkNightSin: with 1 bit, and melted the lions share of my crap
OOC: DarkNightSin: as a fully decked out miner
OOC: DarkNightSin: left me with my jumpsuit and headset basically, not even my pda

@RemieRichards

### LINKED ISSUE #8060: Smoke reagent interactions are not proportionate to the actual amounts.
1 unit sugar, 1 unit synthflesh, 1 unit potassium, and 1 unit phosphorus can heal huge groups of people of over 200 brute AND burn damage almost instantly.

Similarly, replace synthflesh with Fluorosulfuric Acid and it kills hoards of people instantly.

Is this intended?

### LINKED ISSUE #8179: You can inject fluorosulfuric acid into spray bottles with syringes or droppers.
http://i.imgur.com/LOYdHVM.png

Even though it clearly says you shouldn't be able to do this with a beaker transfer.

Should be checked for whatever other chems do this, but it's likely the same problem.

### LINKED ISSUE #8316: Acid Blob Spores create CLOUDS OF ACID on death
BALANCE ISSUE

what the actual fuck, it creates ACID SMOKE CLOUDS when a spore dies

holy fuck thats the most op thing ever

### LINKED ISSUE #8498: Foam Rapidly Replicates Reagents
According to a few players, when your caught in foam you get the full dose of whatever chems were in the beaker every so often, with the net result being that if even one person gets caught in the blast, they'll have much more of the chemical in their blood stream that was actually in the mixture.

### LINKED ISSUE #9216: Metal foam is broken.
Both pre-spawned and hand-made foam grenades only create one piece of foamed metal in the center, surrounded by foam that spreads and vanishes right after.

@Iamgoofball

### LINKED ISSUE #9868: dark matter blob spore clouds OP
We were fighting a blob on sybil and I walked into a cloud of dark matter by a spore dieing in space
Like, 10 lines of red bold text and instant critical :/

### LINKED ISSUE #10107: Explosive gelatin blob spore gas explodes multiple times
While fighting a blob recently, I was struck numerous times with the "exploding tendrils" attack, which put me into crit multiple times. Apparently the "blob tendrils" attacking you actually refers to the gas left by destroyed spores: "The gas the blob spores left exploded every time something touched the center. Including blob spores. A whole bunch rushed over it after someone and it's like b-b-b-b-b-b-b-booom-boom!'

It seems to do a lot of damage, too.

### PR DESCRIPTION
**Pull request by phil235**
- Reagents.reaction() proc and subprocs:
  - fixed the fraction argument in reaction(...., INGESTED, fraction) so the proper amounts of each reagent react, not all of it. (e.g. eating a snack called reaction(..., INGESTED, 1), basically reacting with all reagents instead of only the ones in the bite we take)
  - some optimizations.
  - refactor how protection against touched chems being absorbed is done ( /datum/reagent/proc/reaction_mob(...,TOUCH) was fucked, you could be naked with only a pair of medical shoes and you wouldn't absorb any reagent at all). I created the proc get_permeability_protection() to get that protection. No more simple RNG get touched or not, now the amount that touches you depends on your protection and the volume in question. You now need a full bio suit or equivalent, plus a gas mask or internals on, to be fully protected against the default touch reactions that transfer reagents to your body (but if it's a reagent with its own reaction_mob() proc like acid, you're still fucked). Monkey's mask no longer give it them full protection. The xeno's hard skin give it a decent natural protection. 
  - Adds an argument to the add_reagent() proc to block automatic call of handle_reaction()
  - When using reagents/proc/trans_to(), reactions are now only handled AFTER every reagent is transfered and not before. (fixes foam cleaner grenade containing no space cleaner)
- Smoke/Foam effect changes:
  - All smoke effects now uses process() just like foam.
  - The number of smoke clouds depends on the amount of smoke reagent created.
  - The amount of reagents distributed by the smoke is now never more than the total amount, it is also evenly distributed. Fixes #8060. Fixes #8498. Takes care of half of #8647.
  - fixes not ingesting smoke if it's in the dark.
  - removed BLOCK_GAS_SMOKE_EFFECT flag from clown mask and other non "actual" gas mask.
  - fixes metalfoam not working. Fixes #9216 
  - effect/smoke is made more OOP 
  - Fixes many reagents having effects on mobs not dependent on volume. Fixes #8316. Fixes #9868. 
    Fixes #10107 
  - Remove show_message argument from blob reagent reaction_mob()'s and remove the if(show_message) send_message(). Now send_message is called directly on each blob strike.
- ACID:
  - Damage on humans now depends on acidpwr&volume.
  - changes item/acid_act to take volume into account.  Add a volume check to item/acid_act(), there's a threshold to be able to melt and then a probability to melt (acidpwr&volume). Fixes #6857.
  - When calling human/acid_act(), item acid melting chance lowers after each  items it tries to melt in the list, some of the acid is "used" to melt that item so the next items have a lower chance to melt (to make melting every clothing harder).
  - Sprays can now contain acid again.. Fixes #6276. Fixes #8179.
  - melted storage items now drop their content instead of deleting everything.
- Chem grenade:
  - chem_grenade reaction, if there's no reagent left after reaction (smoke/foam/etc, reagents cleared) you don't get the steam effect and immediate reaction with all atoms around the grenade explosion. (Fixes foam cleaner grenade deleting bloodstains around it as soon as the grenade explodes, even if the foam hasn't reached the tile yet).
  - removing some useless code in chem_grenade/prime().

### CODE REVIEW

**Comment by Iamgoofball** [comment]
:+1:

**Comment by WJohn** [comment]
Jesus christ.

**Comment by Lumipharon** [comment]
Could you give a few examples to show how these changes will behave in practice?
IE: how much smoke reagent you need for X radius
How long it last/how fast it transfers etc.

>The amount of other reagents inside the smoke/foam decides the life expectancy of the effect.
Does that mean if you have just smoke, it's super short lived?

**Comment by phil235** [comment]
>Does that mean if you have just smoke, it's super short lived?
There's a minimum lifetime.

Alright, let's get some examples:
foam radius is minimum 1 (a "plus") and can go up to 20.
Amount of smoke clouds is 1-20.

**Comment by Razharas** [comment]
If only reaction of touch on mob would get a target argument, we could have made real good patches

**Comment by freerealestate** [comment]
That sure is a lot of bullet points I won't read

**Comment by KorPhaeron** [comment]
>removed BLOCK_GAS_SMOKE_EFFECT flag from clown mask and other non "actual" gas mask.

This whole PR is a smokescreen for a stealth clown nerf

**Comment by Iamgoofball** [comment]
@razharas that idea is bad and we've gone over why it's bad

File: code/modules/clothing/masks/gasmask.dm [hunk 1]
```diff
@@ -222,7 +222,7 @@
 /obj/item/clothing/mask/gas/clown_hat
 	name = "clown wig and mask"
 	desc = "A true prankster's facial attire. A clown is incomplete without his wig and mask."
-	flags = MASKCOVERSEYES | BLOCK_GAS_SMOKE_EFFECT | MASKINTERNALS
```
Comment (Jalleo) [inline code/modules/clothing/masks/gasmask.dm]: RIP clown protection against smoke.

**Comment by phil235** [comment]
I still need to do more tests to adjust the different values, maybe tweak the reagents in the hard coded chem grenades, plus add a changelog.

**Comment by optimumtact** [comment]
@phil235 would love to see this soon

**Comment by phil235** [comment]
Alright, should be ready now.

**Pull request merged**